crates/red_knot_test/src/lib.rs

```diff
@@ -65,28 +71,43 @@ fn run_test(test: &parser::MarkdownTest) -> Result<(), Failures> {
         );
         let full_path = workspace_root.join(file.path);
         db.write_file(&full_path, file.code).unwrap();
-        system_paths.push(full_path);
+        paths.push(AbsoluteLineNumberPath {
+            path: full_path,
+            md_offset: file.md_offset,
+        });
     }
 
-    let mut failures = BTreeMap::default();
+    let mut failures = Vec::with_capacity(paths.len());
 
-    for path in system_paths {
-        let file = system_path_to_file(&db, path.clone()).unwrap();
+    for contextual_path in paths {
```

Comment:
naming nit: I don't think `contextual_path` is very clear, I'd name the struct `PathWithOffset` (as mentioned above) and name this `path_with_offset`.